docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,438 @@
+
+The main limitations affecting the current version of Tendermint are the following.
+
+- The current version of the blocksync reactor does not use the full
+  [light client verification](https://github.com/tendermint/tendermint/blob/4743a7ad0/spec/light-client/README.md) algorithm to validate blocks coming from other peers.
+- The code being structured into the blocksync and consensus reactors, only switching from the
+  blocksync reactor to the consensus reactor is supported; switching in the opposite direction is
+  not supported. Alternatively, the consensus reactor could have a mechanism allowing a late node
+  to catch up by skipping calls to `PrepareProposal`/`ProcessProposal`, and
+  `ExtendVote`/`VerifyVoteExtension` and only calling `FinalizeBlock` for each height.
+  Such a mechanism does not exist at the time of writing this RFC.
+
+The blocksync reactor featuring light client verification is being actively worked on (tentatively
+for v0.37). So it is best if this RFC does not try to solve that problem, but just makes sure
+the outcomes are compatible with that effort (TODO: check this paragraph with Callum/Jasmina).
+
+In subsection [Cases to Address](#cases-to-address), we concluded that we can focus on
+solving case (g) in theortical terms.
+However, as the current Tendermint version does not yet support switching back to blocksync once a
+node has switched to consensus, we need to split case (g) into two cases. When a full node needs to
+catch up...
+
+- **(g.1)** ... it has not switched yet from the blocksync reactor to the consensus reactor, or
+
+- **(g.2)** ... it has already switched to the consensus reactor.
+
+This is important in order to discuss the different possible implementations.
+
+#### Base Implementation: Persist and Propagate Extended Commit History
+
+In order to circumvent the fact that we cannot switch from the consensus reactor back to blocksync,
+rather than just keeping the few (TODO: 1?, 2?) most recent extended commits, nodes will need to keep
+and gossip a backlog of extended commits so that the consensus reactor can still propose and decide
+in out-of-date heights.
+
+The base implementation (for which an experimental patch exists) consists in the conservative
+approach of persisting in the block store *all* extended commits for which we have also stored
+the full block.
+Note that this does not apply to light blocks that might have been saved in the block store as a
+result of running statesync at startup.
+
+Then, we modify the blocksync reactor so that peers *always* send requested full blocks together
+with the corresponding extended commit in the `BlockResponse` messages. This guarantees that the
+block store being reconstructed by blocksync has the same information as that of peers that are
+up to date.
+Thus, blocksync has all the data it requires to switch to the consensus reactor, as long as:
+
+- The node is still at height 0 (where no commit or extended commit is needed)
```

Comment:
You're right. And the patch I have in the pipeline takes that into account. Here, though, I wanted to stay simple in the explanations so didn't mention it.
Since you bring it up, that means I probably need to mention it

Reply:
Actually, in those cases, we still call "height 0" the state where we haven't applied any block yet. So if we configure the *initial height* of a chain to be 100. Then the state will evolve like this: "height 0", then "height 100", then "height 101", etc.
I'll try to adapt the wording of the condition based on this.